File: baekjoon/recursive/2529/inequality_sign.py
import sys

def back_tracking(depth):
    global arr, min_, max_

    if depth == LIMIT:
        join = ''.join(arr)
        min_ = min(min_, join)
        max_ = max(max_, join)
        return

    prev_idx = int(arr[depth - 1])
    if signs[depth - 1] == '<':
        start = prev_idx + 1
        end = 10
    else:
        start = 0
        end = prev_idx

    for n in NUMBERS[start:end]:
        if n not in arr[:depth]:
            arr[depth] = n
            back_tracking(depth + 1)


if __name__ == '__main__':
    ZERO = 48
    NUMBERS = ('0', '1', '2', '3', '4', '5', '6', '7', '8', '9')
    def readline(): return sys.stdin.readline()

    LIMIT = int(readline()) + 1
    signs = list(readline().strip().split())
    arr = ['0' for _ in range(LIMIT)]
    min_, max_ = '999999999', '0'
    all = []

    for n in NUMBERS:
        arr[0] = n
        back_tracking(1)

    print(max_, min_, sep='\n')

File: baekjoon/recursive/2529/inequality_sign.cpp
#include <iostream>
#include <algorithm>

int limit;
bool visited[10];
char signs[9];
std::string arr;
std::string max_ = "0";
std::string min_ = "999999999";

void    back_tracking(int depth)
{
    if (depth == limit)
    {
        max_ = std::max(max_, arr);
        min_ = std::min(min_, arr);
        return ;
    }

    char start, end;

    if (signs[depth - 1] == '<')
    {
        start = arr[depth - 1] + 1;
        end = '9';
    }
    else
    {
        start = '0';
        end = arr[depth - 1];
    }

    for (char n = start; n <= end; ++n)
    {
        int idx = n - '0';
        if (!visited[idx])
        {
            arr[depth] = n;
            visited[idx] = true;
            back_tracking(depth + 1);
            visited[idx] = false;
        }
    }
}

int main(void)
{
    std::cin.tie(0)->sync_with_stdio(0);

    std::cin >> limit;

    for (int i = 0; i < limit; ++i)
        std::cin >> signs[i];

    ++limit;
    arr.resize(limit);
    for (char n = '0'; n <= '9'; ++n)
    {
        arr[0] = n;

        int idx = n - '0';
        visited[idx] = true;
        back_tracking(1);
        visited[idx] = false;
    }

    std::cout << max_ << '\n';
    std::cout << min_ << '\n';
    return (0);
}
